syntax = "proto3";

service RaftNode {
  rpc RequestVote(MsgVoteRequest) returns (MsgVoteResponse) {}
  rpc AppendEntries(MsgAppendEntriesRequest) returns (MsgAppendEntriesResponse) {}
}

message MsgVoteRequest {
  uint32 term = 1;
  string candidateId = 2;
  int32 lastLogIndex = 3;
  uint32 lastLogTerm = 5;
}

message MsgVoteResponse {
  uint32 term = 1;
  bool voteGranted = 2;
}

message Entry {
  string key = 1;
  int32 val = 2;
}

message MsgAppendEntriesRequest {
  uint32 term = 1;
  string leaderId = 2;
  int32 prevLogIndex = 3;
  uint32 prevLogTerm = 4;
  uint32 leaderCommit = 5;
  repeated Entry entries = 6;
}

message MsgAppendEntriesResponse {
  uint32 term = 1;
  bool success = 2;
}